include/wil/com.h

```diff
@@ -3002,6 +3002,135 @@ namespace wil
 
 #endif // __IObjectWithSite_INTERFACE_DEFINED__
 
+
+// if C++17 or greater
+#if WIL_HAS_CXX_17
+#ifdef WIL_ENABLE_EXCEPTIONS
+namespace details
+{
+    template<typename>
+    struct com_enumerator_next_traits;
+
+    template<typename Itf, typename T, typename Ret>
+    struct com_enumerator_next_traits<Ret(__stdcall Itf::*)(ULONG, T*, ULONG*)>
+    {
+        using Interface = Itf;
+        using Result = T;
+    };
+
+    template<typename Itf, typename T, typename Ret>
+    struct com_enumerator_next_traits<Ret(__stdcall Itf::*)(ULONG, T*, ULONG*) noexcept>
+    {
+        using Interface = Itf;
+        using Result = T;
+    };
+
+
+    template<typename Interface>
+    struct com_enumerator_traits
+    {
+        using Result = typename com_enumerator_next_traits<decltype(&Interface::Next)>::Result;
+        // If the result is a COM pointer type (IFoo*), then we use wil::com_ptr<IFoo>. Otherwise, we use IFoo*.
+        using smart_result = std::conditional_t<std::is_pointer_v<Result> && std::is_base_of_v<::IUnknown, std::remove_pointer_t<Result>>,
+            wil::com_ptr<std::remove_pointer_t<Result>>, Result>;
+    };
+}
+
+template <typename IEnumType, typename TStoredType = typename details::com_enumerator_traits<IEnumType>::smart_result>
+struct iterator
+{
+    wil::com_ptr<IEnumType> m_enum{};
+    TStoredType m_currentValue{};
+
+    iterator(iterator&&) = default;
+    iterator(iterator const&) = default;
+    iterator& operator=(iterator&&) = default;
+    iterator& operator=(iterator const&) = default;
+
+    iterator(IEnumType* enumPtr) : m_enum(enumPtr)
+    {
+        FetchNext();
+    }
+
+    auto operator->()
+    {
+        return wistd::addressof(m_currentValue);
+    }
+
+    auto& operator*()
+    {
+        return m_currentValue;
+    }
+
+    iterator& operator++()
+    {
+        // If we're already at the end, don't try to advance. Otherwise, use Next to advance.
+        if (m_enum)
+        {
+            FetchNext();
+        }
+
+        return *this;
+    }
+
+    bool operator!=(iterator const& other) const
+    {
+        return !(*this == other);
+    }
+
+    bool operator==(iterator const& other) const
+    {
+        return (m_enum.get() == other.m_enum.get());
+    }
+
+private:
+    void FetchNext()
+    {
+        if (m_enum)
+        {
+            // we cannot say m_currentValue = {} because com_ptr has 2 operator= overloads: one for T* and one for nullptr_t
+            m_currentValue = TStoredType{};
+            auto hr = m_enum->Next(1, &m_currentValue, nullptr);
+            if (hr == S_FALSE)
+            {
+                m_enum = nullptr;
+            }
+            else 
+            {
+                THROW_IF_FAILED_MSG(hr, "Failed to get next");
+            }
+        }
+    }
+};
+
+
+template<typename IEnumXxx, typename TStoredType = typename wil::details::com_enumerator_traits<IEnumXxx>::smart_result>
+WI_NODISCARD auto make_range(IEnumXxx* enumPtr)

```

Comment:
This needs either a new name (so it doesn't conflict with other `make_range` functions, or it needs SFINAE

Reply:
ah got it, that makes sense. the other make_range overloads are:
- make_range(T begin, T end)
- make_range(T begin, size_t count)

This overload is make_range(T), so it shouldn't conflict as long as there isn't another one-param overload. Right?

Reply:
Isn't there `make_range(IVector<T>*)` (& company)?

Even so, I don't think those would conflict since they are more "specific" and would "win" in overload resolution. However, we should still future-proof where we can. Enabling this overload only when the type has `T::Next` (likely of the desired signature) is certainly doable.

Reply:
Nvm, that's `get_range(IVector<T>`, etc. Second comment still stands (and now I'm kind of sad we have two naming conventions here)

Reply:
I'm not sure how you're proposing that we do it while at the same time having a default TStoredType param... did you have a proposal in mind?

Reply:
Is it required to make `TStoredType` a template argument? If not, the function body can deduce this type and it need not be part of the signature. Otherwise, you can do this [the classic SFINAE way](https://godbolt.org/z/881EE5h7x)

Reply:
I've moved TStoredType to not be a template argument